### PR TITLE
Handle persisted queries and expand configuration options

### PR DESCRIPTION
**Pull request by cafreeman**
This PR does a few things:

1. Fixes a bug where we weren't correctly handle array property paths in `processEntities` that caused the client to incorrectly reconstruct the response object when dealing with arrays of entities
2. Updates the client to handle persisted queries by detecting them in the incoming DocumentNode and reformulating the graphql payload to use the [APQ Spec](https://github.com/apollographql/apollo-link-persisted-queries#apollo-engine) for persisted queries.
3. Updates the client config options to allow users to pass in their own fetch instance (rather than passing in an array fetch middleware specifically for `@data-eden/fetch`) as well as a `buildRequest` hook that will allow users to specify how they want the actual HTTP request to be formulated.

### CODE REVIEW

File: internal-packages/graphql-test-server/src/index.ts [hunk 1]
```diff
@@ -1,9 +1,32 @@
...
+import { readJSONSync } from 'fs-extra';
+import path from 'node:path';
+
+const store = readJSONSync(
+  path.resolve(
+    '../react-graphql-test-app/src/__generated__/persisted-documents.json'
+  )
+);
+
+console.log('store', store);
```
Comment (rwjblue) [inline internal-packages/graphql-test-server/src/index.ts:14]: ✂️

File: internal-packages/graphql-test-server/src/index.ts [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
+  plugins: [
+    usePersistedOperations({
+      allowArbitraryOperations: true,
+      getPersistedOperation(hash: string) {
+        console.log('resolving operation for ', hash);
+        return store[hash];
+      },
+    }),
+  ],
+});
```
Comment (rwjblue) [inline internal-packages/graphql-test-server/src/index.ts:29]: Does this mean we always use persisted queries? Or just that we _can_ use them?

File: internal-packages/graphql-test-server/src/schema.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import { createSchema } from 'graphql-yoga';
+import { usePersistedOperations } from '@graphql-yoga/plugin-persisted-operations';
```
Comment (rwjblue) [inline internal-packages/graphql-test-server/src/schema.ts:2]: this imported value seems unused? How is this passing linting?

File: internal-packages/react-graphql-test-app/src/main.tsx [hunk 4]
```diff
@@ -20,6 +21,7 @@ interface Entity {
 export const client = createClient({
   url: 'http://localhost:4000/graphql',
   id: (v: Entity) => `${v.__typename}:${v.id}`,
+  fetch: buildFetch([]),
```
Comment (rwjblue) [inline internal-packages/react-graphql-test-app/src/main.tsx:24]: I think that this use could use the default value instead right? I assume if you don't provide a `fetch` we'll use the native / `globalThis.fetch` right?

File: packages/athena/src/client.ts [hunk 5]
```diff
@@ -16,30 +12,44 @@ import type {
...
 export interface ClientArgs {
   url: string;
   id: IdFetcher;
-  fetchMiddleware?: Array<Middleware>;
+  fetch?: typeof fetch;
+  buildRequest?: BuildRequest;
   adapter: ReactiveAdapter;
 }
 
+export type BuildRequest = (request: GraphQLRequest) => RequestInit;
```
Comment (rwjblue) [inline packages/athena/src/client.ts:25]: Do we want `Request` (as in `new Request(...)`) or `RequestInit` (the second arg to `new Request(url, initHere)`)? Conceptually, I think there are more things that are enabled when providing a whole `Request` object than just returning the `RequestInit`...

Comment (rwjblue) [reply packages/athena/src/client.ts:25]: Additionally, if we just build the request, we can make the URL field optional. 


---

Actually, it's basically more like this:

```ts
interface ClientArgsCommon {
  id: IdFetcher;
  adapter: ReactiveAdapter;
}

interface ClientArgsWithBuildRequest extends ClientArgsCommon {
  buildRequest(GraphQLRequest): Request;
}

interface ClientArgsWithUrl extends ClientArgsCommon {
  url: string;
}

export type ClientArgs = ClientArgsWithUrl | ClientArgsWithBuildRequest;
```

(we need either `url` _or_ `buildRequest`)

File: packages/athena/src/client.ts [hunk 6]
```diff
@@ -62,55 +72,35 @@ export class AthenaClient {
...
-
     try {
-      response = await this.fetch(this.url, {
-        method: 'POST',
-        headers: {
-          'Content-Type': 'application/json',
-        },
-        body: JSON.stringify(body),
-      });
+      response = await this.fetch(this.url, this.buildRequest(request));
```
Comment (rwjblue) [inline packages/athena/src/client.ts:103]: As mentioned above, I'd _generally_ prefer this was:

```suggestion
      response = await this.fetch(this.buildRequest(request));
```

**Review by rwjblue — APPROVED** [review]
Seems good overall!! Just had a few inline questions/comments, but feel free to land at will...

File: internal-packages/graphql-test-server/src/schema.ts (hunk 3 above)
Comment (cafreeman) [reply internal-packages/graphql-test-server/src/schema.ts:2]: it's in `internal-packages` so maybe the linting is different? Either way, I'll remove it

File: packages/athena/src/client.ts (hunk 5 above)
Comment (cafreeman) [reply packages/athena/src/client.ts:25]: huh, yea, good point. I hadn't thought of that but it makes sense. I'll update

Comment (cafreeman) [reply packages/athena/src/client.ts:25]: On further consideration, I'm not totally sure this yields a ton of benefit, but does add some complexity to the actual client constructor. We can always revisit later, but for now I'm going to leave it as is.

**Pull request merged**